Introduce the new menu elements

domfarolino opened the pull request:
This PR introduces the `<menubar>`, `<menulist>`, and `<menuitem>` elements as they appear in Open UI's menu elements proposal: https://open-ui.org/components/menu.explainer/. This includes their integration with popovers, command invokers, `<fieldset>` and checkability, default styles, and variant default styles for the `<menulist>` element depending on how it is rendered. What has not been spec'd yet is:
 - Anchor positioning fallback styles
 - The `orientation` attribute's effect on default styles and keyboard navigability
 - Exact keyboard behavior; we plan to follow customizable select's precedent (see https://github.com/openui/open-ui/issues/1087) of specifying activation behavior, and what happens when you keyboard-navigate _away_ from a menu item (i.e., dismissing `<menulist>` popovers). But to be consistent with customizable select, we're not planning on specifying the exact keyboard behavior for getting from one menulist to another in a menubar, or traversing through menu items in a menu bar or list. These will be covered by examples, notes, and `.optional` WPTs
 - The `checked` event integration discussed in https://github.com/openui/open-ui/issues/1321
 - Figuring out how/if we need to express keyboard type-ahead search (i.e., pressing the `a` key to focus the first menuitem in a menulist that starts with the character `a`)
 - ...

This PR is only a draft. It's suitable for stage 2 evaluation of the menu elements proposal tracked by https://github.com/whatwg/html/issues/11729, but has rough edges and non-trivial TODOs. The API shape is just about fully expressed though.

- [ ] At least two implementers are interested (and none opposed):
   * Chromium
   * [Safari](https://github.com/WebKit/standards-positions/issues/580)
   * [Firefox](https://github.com/mozilla/standards-positions/issues/1317)
- [ ] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * (in progress) https://github.com/web-platform-tests/wpt/tree/master/html/semantics/menu/tentative
- [ ] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://issues.chromium.org/issues/406566432
   * Gecko: …
   * WebKit: …
   * Deno (only for timers, structured clone, base64 utils, channel messaging, module resolution, web workers, and web storage): N/A
   * Node.js (only for timers, structured clone, base64 utils, channel messaging, and module resolution): N/A
- [ ] Corresponding [HTML AAM](https://w3c.github.io/html-aam/) & [ARIA in HTML](https://w3c.github.io/html-aria/) issues & PRs:
- [ ] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: …
- [ ] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/12011/dom.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">/dom.html</a>  ( <a href="https://whatpr.org/html/12011/68909b2...b37a5a8/dom.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">diff</a> )
<a href="https://whatpr.org/html/12011/form-elements.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">/form-elements.html</a>  ( <a href="https://whatpr.org/html/12011/68909b2...b37a5a8/form-elements.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">diff</a> )
<a href="https://whatpr.org/html/12011/grouping-content.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">/grouping-content.html</a>  ( <a href="https://whatpr.org/html/12011/68909b2...b37a5a8/grouping-content.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">diff</a> )
<a href="https://whatpr.org/html/12011/index.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">/index.html</a>  ( <a href="https://whatpr.org/html/12011/68909b2...b37a5a8/index.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">diff</a> )
<a href="https://whatpr.org/html/12011/indices.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">/indices.html</a>  ( <a href="https://whatpr.org/html/12011/68909b2...b37a5a8/indices.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">diff</a> )
<a href="https://whatpr.org/html/12011/infrastructure.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/12011/68909b2...b37a5a8/infrastructure.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">diff</a> )
<a href="https://whatpr.org/html/12011/interaction.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">/interaction.html</a>  ( <a href="https://whatpr.org/html/12011/68909b2...b37a5a8/interaction.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">diff</a> )
<a href="https://whatpr.org/html/12011/interactive-elements.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">/interactive-elements.html</a>  ( <a href="https://whatpr.org/html/12011/68909b2...b37a5a8/interactive-elements.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">diff</a> )
<a href="https://whatpr.org/html/12011/obsolete.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">/obsolete.html</a>  ( <a href="https://whatpr.org/html/12011/68909b2...b37a5a8/obsolete.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">diff</a> )
<a href="https://whatpr.org/html/12011/popover.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">/popover.html</a>  ( <a href="https://whatpr.org/html/12011/68909b2...b37a5a8/popover.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">diff</a> )
<a href="https://whatpr.org/html/12011/rendering.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">/rendering.html</a>  ( <a href="https://whatpr.org/html/12011/68909b2...b37a5a8/rendering.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">diff</a> )
<a href="https://whatpr.org/html/12011/semantics-other.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">/semantics-other.html</a>  ( <a href="https://whatpr.org/html/12011/68909b2...b37a5a8/semantics-other.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">diff</a> )
<a href="https://whatpr.org/html/12011/text-level-semantics.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">/text-level-semantics.html</a>  ( <a href="https://whatpr.org/html/12011/68909b2...b37a5a8/text-level-semantics.html" title="Last updated on Aug 6, 2026, 7:00 PM UTC (b37a5a8)">diff</a> )

domfarolino committed:
HTMLMenuBarElement

domfarolino committed:
HTMLMenuListElement basics and popover generalization

domfarolino committed:
Complete popover integration, with open questions

Note there are a couple open questions in the form of TODOs. For example:

1. Should `popovertargetaction` be allowed to target `<menulist>` elements?
   Probably, but it could use discussion.
2. Should the attribute change steps that react to `popover` attribute
   mutations "work" on `<menulist>` elements? This effectively makes the
   explicit `popover` attribute work on `<menulist>` elements, possibly
   overriding their default popover behavior. This could also use more
   discussion. (My preference would be to ignore `popover` on `<menulist>`
   for now).

domfarolino committed:
Move command/commandfor attrs into their own section

This commit moves the command/commandfor attributes into their own section,
since they're general to both buttons and menuitem elements. The next steps are:

 1. Pull out a shared activation behavior algorithm from the button element,
 that defines the command-invoking parts that the two elements have in common
 2. Add the three new commands
 3. Define the menuitem element, and any menuitem-specific things for the
    commands

domfarolino committed:
More menulist default styles

domfarolino committed:
Checkedness and fieldset wired up

domfarolino committed:
Refactor checkedness for activation path

domfarolino committed:
Checkmark styles

domfarolino committed:
More conditional default styles

domfarolino committed:
Define menulist command steps

These are copied from the main button activation steps's popover handling steps

domfarolino committed:
Break out the shared command invoker activation steps

domfarolino committed:
Oops fix build

mtrootyy commented on source:
```diff
@@ -21604,6 +21647,451 @@ interface <dfn interface>HTMLMenuElement</dfn> : <span>HTMLElement</span> {
+   <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
+   <dd>Zero or <span><code>menuitem</code> element inner content elements</span>.</dd>
+   <dt><span data-x="concept-element-attributes">Content attributes</span>:</dt>
+   <dd><span>Global attributes</span></dd>
```
The "command" & "commandfor" attribute should be added .

mtrootyy commented on source:
```diff
@@ -21604,6 +21647,451 @@ interface <dfn interface>HTMLMenuElement</dfn> : <span>HTMLElement</span> {
+  <dl class="element">
+   <dt><span data-x="concept-element-categories">Categories</span>:</dt>
+   <dd><span><code>menulist</code> element inner content elements</span>.</dd>
+   <dd>If the element's children include at least one <code>menuitem</code> element: <span>Palpable content</span>.</dd>
```
Can menuitem element be placed inside menuitem element?

mtrootyy commented on source:
```diff
@@ -148307,6 +148958,110 @@ select optgroup legend {
+  padding-block: 0;
+  padding-inline: 0;
+}
+
```
I think that it seems necessary to add CSS of "menulist > fieldset > legend".

mtrootyy commented on source:
```diff
@@ -148307,6 +148958,110 @@ select optgroup legend {
+}
+
+menulist {
+  display: flex;
```
If the menulist element is exactly a "native" popover, shouldn't the default of initial state of the menulist element be "display:none"?

dbaron replied:
The latter rule is correct -- exactly that rule is in Blink's UA stylesheet.  I've added it.

mtrootyy commented:
In section of "4.10.16 The legend element";
- "Content model:" of "if the element is a descendant of a menulist element" should be clarified.

mtrootyy commented on source:
```diff
@@ -148307,6 +148958,110 @@ select optgroup legend {
+  border-radius: 0.25em;
+  padding: 0.25em;
+}
+
```
The following should be added;

```
menubar[orientation="vertical"]  {
  flex-direction: column;
}
```

dbaron committed:
Move the new menu elements to the interactive elements section.

(This is a text move with no other changes.)

dbaron committed:
Minor fixes to palpable content and * inner content elements definitions.

dbaron committed:
Limit invoked submenu definition to only toggle-menu command.

This limits the invoked submenu definition to only the toggle-menu
command.  It previously included the show-menu and hide-menu commands as
well.

The various UA-implemented behaviors that result from being an invoked
submenu involve both opening and closing menus.  This avoids the
weirdness of either (a) having those UA behaviors close menus as a
result of a show-menu command or open menus as a result of a hide-menu
command or (b) having to add a lot of additional complexity to those
behaviors (that I don't think anyone actually wants) in order to
correctly differentiate between the three commands.

dbaron committed:
Hook up "menulist element inner content elements" definition.

The "menulist element inner content elements" definition was defined,
but not previously used for the content model of the menulist element.

dbaron committed:
Allow <fieldset> in <menubar>.

This allows <fieldset> in <menubar> per the discussion in
https://github.com/openui/open-ui/issues/1433

dbaron committed:
Allow div/span within menubar/menulist/menuitem.

This adds allowance for div and span within menubar and menulist while
propagating the content model restrictions to their descendants.

Previously div and span were allowed in menuitem without propagating the
content model restrictions; this now propagates the restrictions
correctly.

dbaron committed:
Add styles and note about application/navigation menu distinction.

This adds styles to hide links (as has long been documented in the
explainer) to discourage use of these new menu elements for navigation
menus.

This also adds an addition note describing the distinction and linking
to APG (which seems like the best authoritative-looking source that I
can find, although I wouldn't mind a better one).

dbaron committed:
Hide links in menubar/menulist, not just menuitem.

Followup to commit 7e36c17b0c352471d82070f57278378d2f3d02e3.

dbaron committed:
Add activation steps for checkable menu items.

This adds the steps for checking/unchecking the menu items and firing
the checked event as resolved in
https://github.com/openui/open-ui/issues/1321

It also removes a duplicate item in a definition (although it's possible
that something non-duplicate was intended, but I don't know what!).

dbaron committed:
Remove show-menu and hide-menu commands.

Per discussion in https://github.com/openui/open-ui/issues/1442

dbaron committed:
List command and commandfor attributes on menuitem.

dbaron committed:
Make closed menulists display:none.

This is as pointed out in
https://github.com/whatwg/html/pull/12011#discussion_r2616912648
and as already present in Blink's UA stylesheet.

dbaron committed:
Propagate content model through <fieldset>, and stop trying (not correctly) to allow <span> around <menuitem>.

mtrootyy commented on source:
```diff
@@ -22630,6 +22736,18 @@ included with Exhibit B.
    <span><code>select</code> element inner content elements</span>.</dd>
+   <dd>Otherwise, if the element is a descendant of a <code>fieldset</code> that is a descendant of
+   a <code>menubar</code> element: Zero or more <span><code>menubar</code> element
+   <code>fieldset</code> element inner content elements</span>.</dd>
```
Because it can be specified as follows, there is no need to create new categories such as "menubar element fieldset element inner content elements".
> if the element is a descendant of a fieldset that is a descendant of a menubar element: Zero or more menubar element inner content elements, except fieldset elements.

dbaron committed:
Update to nested DOM structure for menus and drop toggle-menu command.

dbaron committed:
Change invalid value default of <fieldset checkable> from None to Multiple.

dbaron committed:
Remove the menubar orientation attribute, at least for now.

dbaron committed:
Fix case of HTMLSubMenuElement interface.

lukewarlow commented on source:
```diff
@@ -58882,6 +58777,78 @@ interface <dfn interface>HTMLFieldSetElement</dfn> : <span>HTMLElement</span> {
+  value default">missing value default</i> and <i data-x="empty value default">empty value
+  default</i> are the <dfn data-x="attr-fieldset-checkable-state-none">None</dfn> state. This
+  attribute's <i data-x="invalid value default">invalid value default</i> is the <span
+  data-x="attr-fieldset-checkable-state-multiple">Multiple</span> state.</p>
```
I'm not sure this makes sense as the invalid value default? I would expect none?

dbaron replied:
I'm curious why you think invalid should go to none rather than match the empty value default?  (Now that I've made it do so, see my other comment.)

(I think what I've ended up with now makes the reflection and defaulting similar to [the `crossorigin` attribute](https://html.spec.whatwg.org/multipage/urls-and-fetching.html#cors-settings-attribute).)

lukewarlow commented on source:
```diff
@@ -58882,6 +58777,78 @@ interface <dfn interface>HTMLFieldSetElement</dfn> : <span>HTMLElement</span> {
+  data-x="attr-fieldset-checkable">checkable</code></dfn> attribute is an <span>enumerated
+  attribute</span> with the following keywords and states. This attribute's <i data-x="missing
+  value default">missing value default</i> and <i data-x="empty value default">empty value
+  default</i> are the <dfn data-x="attr-fieldset-checkable-state-none">None</dfn> state. This
```
Perhaps for empty value default single makes more sense. Though Im not super sure either way.

dbaron replied:
I initially wanted to make the empty and invalid value defaults match.  Then I thought the reflection was forcing me to make the empty and missing value defaults match, until I realized that reflecting as `DOMString?` was a better solution, and I could make the empty and invalid value defaults match.  I've now updated both to be multiple, since I think multiple is the simpler behavior (whereas single does more and enforces exclusivity).

lukewarlow replied:
Yeah thinking on it some more I think multiple makes sense.

dbaron committed:
Reflect <fieldset checkable> attribute, modeled after <link crossorigin> reflection.

mtrootyy commented on source:
add;
> As chiled of submenu element.

dbaron replied:
Fixed in bbde238d54c69f815cec22c8917a08fa27c1074b.

mtrootyy commented on source:
add;
> As chiled of submenu element.

dbaron replied:
Fixed in bbde238d54c69f815cec22c8917a08fa27c1074b.

dbaron committed:
Add "as a child of a submenu element" to contexts.

I missed this in 6a55bd869e0b70e779a9a3aca8844f15718c134a as @mtrootyy
pointed out.

dbaron committed:
Specify click-retargeting behavior for menuitem activation.

This has become easier to do following the content model changes, and
specifies the behavior that was desired in
https://github.com/openui/open-ui/issues/1312

dbaron committed:
Specify :open pseudo-class for menuitem elements.

This is as resolved in https://github.com/openui/open-ui/issues/1363

dbaron committed:
Add key handling for menuitems, part 1.

(This adds many TODOs that I need to fix in followups.)

dbaron committed:
Add key handling for menuitems, part 2.

(Still more to come.)

dbaron committed:
Add key handling for menuitems, part 3.

(Still more to come.)

This also fills in some definitions that were needed and fixes one
inconsistent set of definitions.

dbaron committed:
Add key handling for menuitems, part 4 of 4.

dbaron committed:
Add another TODO and clarify some of the existing ones.

dbaron committed:
Fix TODO in activation behavior by hooking it up to checking algorithm.

This fixes the TODO about closing the containing menulist in the
menuitem activation behavior by hooking that algorithm up to the
checking algorithm, which reports whether to close the containing
menulist.

This also changes the menuitem's concept of checkability from a boolean
to exposing the tri-state attribute state, since that's more useful.

This also moves firing of the checked event into the checking algorithm
(without otherwise changing its behavior, but see added TODOs).

This does change the closing behavior slightly: this now closes the
containing menulist when unchecking a radio menuitem, just like when
checking it.  This is what is implemented in Chrome and I think the
distinction in the spec was likely inadvertent.

This also adds quite a few TODOs for additional bugs that I found while
doing this but didn't fix.

dbaron committed:
Add note about firing events matching checkbox but not radio.

dbaron committed:
Specify some aspects of pointer/drag opening of submenus.

As discussed in
https://github.com/openui/open-ui/issues/1190#issuecomment-3779238202

smaug---- commented:
(I'll continue looking into this some more, but some generic feedback already)

Overall the new approach seems a lot easier to understand and event handling works in a more natural way etc.

I'm still not a big fan of fieldset and checkability handling. Makes it rather awkward to create menus like
<img width="303" height="227" alt="image" src="https://github.com/user-attachments/assets/88d78fc2-7e12-490f-bc72-2e008c25541f" />
and often the fieldset in menus seems to be just useless. submenu itself does quite a bit grouping, underneath that name attribute might work fine, if needed (usually it wouldn't be needed).

lukewarlow commented on source:
This definitely feels like a missing piece. We should I think apply `position-area: block-end span-inline-end;` when a buttons opened the menulist.
